### ISSUE TITLE
DisplayPort MST

### ISSUE THREAD

**Issue by ascent12** [SirCmpwn/wlroots#130]
This is a tracking issue for DisplayPort MST (Multi-Stream Transport), which is the ability for multiple monitors to come off the same DisplayPort connector. It's also used in Dual-panel 4K monitors and many modern laptop docks.

Unfortunately, I don't have a lot of the hardware to implement this all properly or even be certain about the requirements for it. I DO own a monitor that is MST compatible, but the radeon drivers are messed up with my GPU, so I can't get it working.

Based on my understanding (I may be wrong), what we need to do is
- [x] Handle DRM connectors being able to appear at any point
- [ ] Handle DRM connectors being able to disappear at any point
- [ ] Naming the connectors properly (e.g. DP-1-1, DP-1-2)
- [ ] Being able to handle the DRM tiling property (for dual-panel 4K screens), possibly exposing dual-panels as a single logical monitor?

**Comment by SirCmpwn** [comment]
Does daisy chaning need special consideration?

**Comment by SirCmpwn** [comment]
Also, it might be worthwhile to put down a hardware wishlist of some cheap hardware with these capabilities that someone interested in support for these features might donate to you.

**Comment by ascent12** [comment]
I believe daisy chaining works the same as a MST hub. My bought a Dell U2715H recently, which is daisy chaining compatible, but the problems are with my GPU (HD 7950) and its drivers now. Radeon SHOULD support MST, but it's currently experimental, and doesn't currently work properly for me.

The only other GPUs I have on hand is some old Nvidia (570?) card and an AMD HD 6950. It doesn't have a displayport, and I don't even have a clue about the status of MST for nouveau anyway.

Once the AMDGPU DC stuff lands in the kernel (this may be several releases away?), I may pick up a new cheapish AMD GPU, and that should be able to handle MST properly.

**Comment by kasicka** [comment]
I have a few thinkpads/laptops laying around at home if needed, but it's not bleeding edge hardware.

**Comment by ascent12** [comment]
Regarding naming the connectors, the DRM system seems to handle this pretty well for us. It'll just keep counting on them like other DisplayPort connectors (DP-2, DP3, ...).
There are some weird cases where it doesn't un-number the old connectors when they get unplugged, meaning when I plugged it back in, I got DP-5 and DP-6, when I really should have gotten DP-3 and DP-4. I'm not 100% how to reproduce it, though.

X.org names the connectors themselves in a special way, using the DRM PATH connector property, giving names like "DisplayPort-2-1" and "DisplayPort-2-8". I don't know if we want to put the extra effort in to have a similar naming scheme or not.
